### DISCUSSION TITLE
Why no `Impl Iterator for StreamingPeekableIter` in `git-packetline`?

### DISCUSSION THREAD

**Discussion by ethever** [GitoxideLabs/gitoxide#521] [Q&A] [answered]
currently the git_packetline crate streamingly reads packetline from `Read`, and it provides git_packetline::StreamingPeekableIter::peek_line and git_packetline::StreamingPeekableIter::read_line.

impl Interator for StreamingPeekableIter. This will be helpful for iterating packetline over an streming `Read`.

the method `lines` exists for struct `StreamingPeekableIter<BufReader<&[u8]>>`, but its trait bounds were not satisfied
the following trait bounds were not satisfied:
`StreamingPeekableIter<BufReader<&[u8]>>: BufRead`
which is required by `&mut StreamingPeekableIter<BufReader<&[u8]>>: BufRead`.

**Comment by Byron** [comment, marked as answer]
_(I have edited this former issue to better fit into this Q&A format, merging comments with content and adding some punctuation)_

Believe me, I tried, but ultimately self-referencing iterators are not possible without GATs. That said, once GATs are more common this should officially become a streaming iterator.

**Reply by ethever** [reply]
I got it, appreciate for your reply.
